### ISSUE TITLE
Typescript imports don't find `index.d.ts` when importing individual components

### ISSUE THREAD

**Issue by Renari** [bootstrap-vue/bootstrap-vue#1976]
- Version of Bootstrap-Vue that you are using
2.0.0-rc.11

```js
import bContainer from 'bootstrap-vue/es/components/layout/container';
```

```
69:24 Could not find a declaration file for module 'bootstrap-vue/es/components/layout/container'. 
'/node_modules/bootstrap-vue/es/components/layout/container.js' implicitly has an 'any' type.
  Try `npm install @types/bootstrap-vue` if it exists or add a new declaration (.d.ts) file containing `declare module 'bootst
rap-vue';`
  > 69 | import bContainer from 'bootstrap-vue/es/components/layout/container';
       |                        ^
```

I'm not sure if TypeScript has a way to specify the type declarations. From what I've looked into it doesn't. So to support imports like this a TypeScript definition would need to be made in every component directory.

**Comment by RehanSaeed** [comment]
Can confirm that this is an issue.

**Comment by MartinYounghoonKim** [comment]
Hi there
is it solved ?

**Comment by nicolasigot** [comment]
Having the same issue and noticed that ScrollSpy component seems to be  missing from definition when trying to import it as a plugin:
import { Scrollspy } from 'bootstrap-vue/es/directives';
Any idea ?

**Comment by Renari** [comment]
@nicolasigot You can create your own definitions in your `shims-vue.d.ts` file.

e.g.
```declare module 'bootstrap-vue/es/components/layout/container' {
  import Vue from 'vue';
  export default Vue;
}```

**Comment by essenmitsosse** [comment]
This actually works, which is great. But I can only import one component per .d.ts file. 

If I try to import multiple components, it either complains, about importing 'vue' multiple times:
```typescript
import Vue from 'vue';

declare module 'bootstrap-vue/es/components/button/button' {
	export default Vue;
}

declare module 'bootstrap-vue/src/components/layout/col' {
	export default Vue;
}
```

if I globally import 'vue' in the file:
```typescript
import Vue from 'vue';

declare module 'bootstrap-vue/es/components/button/button' {
	export default Vue;
}

declare module 'bootstrap-vue/src/components/layout/col' {
	export default Vue;
}
```

`Invalid module name in augmentation. Module 'bootstrap-vue/es/components/button/button' resolves to an untyped module at '/Volumes/git/code/app-prokoo/node_modules/bootstrap-vue/es/components/button/button.js', which cannot be augmented.`

Is there a way to import all bootstrap-vue typings, but only import individual components into the build?

**Comment by yooouuri** [comment]
Some progress on this issue?

**Comment by idrislab** [comment]
any update on this issue?

**Comment by tmorehouse** [comment]
PR #3209 may address this issue. I'm not an expert at creating type declaration files, and was unable to find anyone to assist/test fully (and no one raising issues re typescript volunteered to help), so I can't guarantee it will work.

**Comment by soyersoyer** [comment]
It's not working out of the box with the new plugins, the warning is:
```
Could not find a declaration file for module 'bootstrap-vue/es/components'. '/web/node_modules/bootstrap-vue/es/components/index.js' implicitly has an 'any' type.
  Try `npm install @types/bootstrap-vue` if it exists or add a new declaration (.d.ts) file containing `declare module 'bootstrap-vue/es/components';`
```

But its easier now, because I have to create only one d.ts file:
```
declare module 'bootstrap-vue/es/components' {
    import Vue, { PluginFunction, PluginObject } from 'vue';

    class BootstrapVue implements PluginObject<{}> {
        [key: string]: any,
        public install: PluginFunction<{}>
    }

    export const ModalPlugin: BootstrapVue;
    export const NavbarPlugin: BootstrapVue;
    export const CollapsePlugin: BootstrapVue;
    export const BreadcrumbPlugin: BootstrapVue;
    export const FormTextareaPlugin: BootstrapVue;
    export const FormFilePlugin: BootstrapVue;
    export const ToastPlugin: BootstrapVue;
}
```

**Comment by tmorehouse** [comment]
@soyersoyer could you create a new issue, and maybe we can get some improvements to the new types implemented.

**Comment by tmorehouse** [comment]
@soyersoyer what if you tried creating the following d.ts file:

```ts
declare module 'bootstrap-vue/es/components' {
  export * from 'node_modules/bootstrap-vue/types/components'
}

declare module 'bootstrap-vue/es/directives' {
  export * from 'node_modules/bootstrap-vue/types/directives'
}

declare module 'bootstrap-vue/es/bv-config' {
  export * from 'node_modules/bootstrap-vue/types/bv-config'
}

declare module 'bootstrap-vue/es' {
  import { BvPlugin, BvConfigOptions } from './bv-plugin'

  export * from 'node_modules/bootstrap-vue/types/bv-config'
  export * from 'node_modules/bootstrap-vue/types/bv-event'
  export * from 'node_modules/bootstrap-vue/types/components'
  export * from 'node_modules/bootstrap-vue/types/directives'

  declare const BootstrapVue: BootstrapVue
  export default BootstrapVue
  export interface BootstrapVue extends BvPlugin {
    setConfig: (config: BvConfigOptions) => void
  }
}
```

I am curious to see if this works, and if it does, we'll include it in our types declaration in the next release.

**Comment by soyersoyer** [comment]
Thanks for your help.

I fixed the import path:
 
```import { BvPlugin, BvConfigOptions } from './bv-plugin'``` ->  ```import { BvPlugin, BvConfigOptions } from 'node_modules/bootstrap-vue/types/bv-plugin'```

But there are some errors:

 ```declare const BootstrapVue: BootstrapVue``` -> ```A 'declare' modifier cannot be used in an already ambient context.```

```Vue.use(ToastPlugin);``` -> ```'ToastPlugin' only refers to a type, but is being used as a value here.```

You can easily setup a working vue typescript environment with vue cli. It's less than 1 minute! :)

**Comment by soyersoyer** [comment]
With
```
export const BootstrapVue: BootstrapVue
//export default BootstrapVue
```
And the same with the plugins:
```
export const ModalPlugin: ModalPlugin
//export default ModalPlugin
```
seems to work.

**Comment by tmorehouse** [comment]
@soyersoyer The following _might_ work better:

```ts
declare module 'bootstrap-vue/es/components' {
  import { Component, VueConstructor } from 'vue'
  import { BvPlugin, BvConfigOptions } from 'node_modaules/bootstrap-vue/types/bv-plugin'
  
  const BVComponents: {
    default: Component & BvPlugin & VueConstructor
    [key: string]: Component & BvPlugin & VueConstructor
  }

  export = BVComponents
}

declare module 'bootstrap-vue/es/components/*' {
  import { Component, VueConstructor } from 'vue'
  import { BvPlugin, BvConfigOptions } from 'node_modaules/bootstrap-vue/types/bv-plugin'
  
  const BVComponents: {
    default: Component & BvPlugin & VueConstructor
    [key: string]: Component & BvPlugin & VueConstructor
  }

  export = BVComponents
}

declare module 'bootstrap-vue/es/directives' {
  import { VueConstructor, DirectiveOptions } from 'vue'
  import { BvPlugin, BvConfigOptions } from 'node_modaules/bootstrap-vue/types/bv-plugin'

  const BVDirectives: {
    default: DirectiveOptions & BvPlugin & VueConstructor
    [key: string]: DirectiveOptions & BvPlugin & VueConstructor
  }

  export = BVDirectives
}

declare module 'bootstrap-vue/es/directives/*' {
  import { VueConstructor, DirectiveOptions } from 'vue'
  import { BvPlugin, BvConfigOptions } from 'node_modaules/bootstrap-vue/types/bv-plugin'

  const BVDirectives: {
    default: DirectiveOptions & BvPlugin & VueConstructor
    [key: string]: DirectiveOptions & BvPlugin & VueConstructor
  }

  export = BVDirectives
}


declare module 'bootstrap-vue/es/bv-config' {
  import BVConfigPlugin from 'node_modaules/bootstrap-vue/types/bv-config'

  export default BVConfigPlugin
}

declare module 'bootstrap-vue/es' {
  import BootstrapVue from 'bootstrap-vue'

  export default BootstrapVue
}
```

**Comment by tmorehouse** [comment]
It appears that Typescript has "issues" with resolving sub-modules, and not checking the `package.json` 'types' to load in the sub-module declarations (it ignores the main packages package.json file.

We have updated the type declarations to reside in the `/src/` directories, next to the individual module index.js files, and copy these files over to the `/es/` directories during build.  This should be available in the upcoming 2.0.0-rc.21 release.

**Comment by tmorehouse** [comment]
Version 2.0.0-rc.21 has been released

If you run into any issues with the new types, please open a new issue.
